lms/services/blackboard_api/_basic.py

```diff
@@ -63,28 +61,13 @@ def __init__(
 
         self._http_service = http_service
         self._oauth_http_service = oauth_http_service
-        self._oauth2_token_service = oauth2_token_service
 
     def get_token(self, authorization_code):
-        # Send a request to Blackboard to get an access token.
-        response = self._http_service.post(
-            self._api_url("oauth2/token"),
-            data={
-                "grant_type": "authorization_code",
-                "redirect_uri": self.redirect_uri,
-                "code": authorization_code,
-            },
+        self._oauth_http_service.access_token_request(
+            token_url=self._api_url("oauth2/token"),
+            redirect_uri=self.redirect_uri,
```

Comment:
Passing these two values every time strikes me as a bit odd. If this wasn't a service, I'd expect the interface to be something more like:

```
client = OAuthThing(token_url=..., redirect_uri=..., client_id=..., client_secret=...)
client.access_token_request()
client.refresh_token_request(token)
```

These don't seem like things that change over the life-time of a non-service object. This would mean less state having to be held by the caller which is only of interest to the object. Basically there's a lack of encapsulation here I think.

This suggests to me that perhaps what we want from the service is for it to return an object we can make requests with, rather than be the object we make requests with. So maybe something like:

```
client = self._oauth_http_service.get_authenticated_session(token_url=..., redirect_uri=..., client_id=..., client_secret=...)
```

Reply:
Yeah, I think having a two-step process where you first create an `OAuthClient` object (passing it reusable settings like `client_id`, `client_secret`, etc) and then make a request (passing only request-specific arguments) likely makes sense. I had a scratch version of that at one point but I held off on doing it right now because I wanted to minimize the scope of last-second code changes in order to get the feature shipped on time.

I don't think this is a question of service or not-service though. It's useful for `OAuthHTTPService` to be a service because it relies on two other services (`HTTPService` and `OAuth2TokenService`) and the service factory can inject those, hiding those dependencies from `OAuthHTTPService`'s users. In future the service can also inject any other services or config settings etc that `OAuthHTTPService` comes to depend on or can vary how it constructs the service depending on the context etc, in general hiding service construction from the service's users is useful.

So I think it's a question of how `OAuthHTTPService` is designed.

The scratch version I had was something like this:

```python

Reply:
> This suggests to me that perhaps what we want from the service is for it to return an object we can make requests with, rather than be the object we make requests with. So maybe something like:
> 
> ```
> client = self._oauth_http_service.get_authenticated_session(token_url=..., redirect_uri=..., client_id=..., client_secret=...)
> ```

Ah, yes pretty much exactly what I was thinking. But I'm not totally sure. `get_authenticated_session()` is going to end up with too many arguments.

Reply:
Yeah, I was just suggesting I suspect you ended up here because you started with a service, rather than you can't do this pattern with one. Yay factory factory?

Also if X is too many arguments on the constructor, better that than X + one or two per call?

Reply:
Just had a brainwave: You could combine both to get the best of both. So have a `OAuthSettingsBag`, and then pass it through to the factory factory method.

Reply:
Yes, perhaps.

I didn't end up here because I started with a service. I actually wrote the `OAuthService.make_client()` version but it wasn't a clear win to me for various reasons. It still has the too-many-arguments problem (if you don't add a `SettingsBag` class) and it also makes user code and the tests for that user code more complicated. So I just backed off from it for now. But I think we could revive at some point when there's more time to consider.